urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/nn/Links.java

```diff
@@ -0,0 +1,125 @@
+ * Copyright 2012 Yuriy Lagodiuk
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ******************************************************************************/
+package com.lagodiuk.nn;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlTransient;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import com.lagodiuk.nn.serializing.xml.MapAdapter;
+
+public class Links implements Cloneable {
+
+	@XmlJavaTypeAdapter(value = MapAdapter.class)
+	private Map<Integer, Map<Integer, Double>> links = new LinkedHashMap<Integer, Map<Integer, Double>>();
+
+	@XmlElement(name = "linksCount")
+	private int totalLinksCount = 0;
+
+	public Collection<Integer> getReceivers(int activatorNeuronNumber) {
+		Collection<Integer> ret = null;
+		if (this.links.containsKey(activatorNeuronNumber)) {
+			ret = Collections.unmodifiableSet(this.links.get(activatorNeuronNumber).keySet());
+		} else {
+			ret = Collections.emptySet();
+		}
+		return ret;
+	}
+
+	public Double getWeight(int activatorNeuronNumber, int receiverNeuronNumber) {
+		double weight = 0;
+
+		if (this.links.containsKey(activatorNeuronNumber)) {
+			Map<Integer, Double> receiverNumToWeight = this.links.get(activatorNeuronNumber);
+
+			if (receiverNumToWeight.containsKey(receiverNeuronNumber)) {
+				weight = receiverNumToWeight.get(receiverNeuronNumber);
+			} else {
+				throw new IllegalArgumentException();
+			}
+		} else {
+			throw new IllegalArgumentException();
+		}
+		return weight;
+	}
+
+	public void addWeight(int activatorNeuronNumber, int receiverNeuronNumber, double weight) {
+		if (!this.links.containsKey(activatorNeuronNumber)) {
+			this.links.put(activatorNeuronNumber, new LinkedHashMap<Integer, Double>());
+		}
+		this.links.get(activatorNeuronNumber).put(receiverNeuronNumber, weight);
+		this.totalLinksCount++;
+	}
+
+	@XmlTransient
+	public List<Double> getAllWeights() {
+		List<Double> weights = new ArrayList<Double>(this.totalLinksCount);
+
+		for (Integer activatorIndx : this.links.keySet()) {
+			Map<Integer, Double> receiverIndxToWeight = this.links.get(activatorIndx);
+
+			for (Integer receiverIndx : receiverIndxToWeight.keySet()) {
+				weights.add(receiverIndxToWeight.get(receiverIndx));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.